python/sdist/amici/de_model.py

```diff
@@ -509,6 +501,26 @@ def get_rate(symbol: sp.Symbol):
             self._syms["w"] = sp.Matrix(topo_expr_syms)
             self._eqs["w"] = sp.Matrix(list(w_sorted.values()))
 
+        # replace rateOf-instances in x0 by xdot equation
+        made_substitutions = False
+        for i_state in range(len(self.eq("x0"))):
+            new, replacement = self._eqs["x0"][i_state].replace(
+                rate_of_func, get_rate, map=True
+            )
+            if replacement:
+                self._eqs["x0"][i_state] = new
+                made_substitutions = True
+        if made_substitutions:
+            # Replace any newly introduced state variables
+            #  by their x0 expressions.
+            # Also replace any newly introduced `w` symbols by their
+            #  expressions (after `w` was toposorted above).
+            subs = toposort_symbols(
+                dict(zip(self.sym("x_rdata"), self.eq("x0"), strict=True))
+            )
+            subs = dict(zip(self._syms["w"], self.eq("w"), strict=True)) | subs
```

Comment:
I think this has to be done in a loop as `dict(zip(self._syms["w"], self.eq("w"), strict=True))` subs may introduce new `self.sym("x_rdata")` variables.

Reply:
`w` should not depend on `x_rdata`:

https://github.com/AMICI-dev/AMICI/blob/e38afde6423847698380158778464aa423963347/python/sdist/amici/_codegen/cxx_functions.py#L313-L319

Unless the respective substitutions in `w` are only made later, but I don't think that is the case -  the rateOf processing should happen last.